Alternative: Lookup temperature for broader range (from -200 to +800 C°)

rudihorn opened the pull request:
Alterantive PR for #2

RCasatta committed:
remove compiler flag preventing the build

RCasatta committed:
implement temperature lookup for broader range including negative values

Commit:
refactor lots of code

Commit:
make pub

RCasatta commented:
thanks for your work...

>  I still want to at the very least upgrade the stm32 library reference to the newer stm32f1xx-hal before submitting a new version to crates and still need to check all the examples work

Totally agree some dep updates is required before the new version, I can test examples with pt1000 on stm32f1xx-hal and also with stm32f4xx-hal if needed

How do you imagine the PT1000 use in the final API, an enum parameter added to `read_default_conversion` that use the relative table?

I guess `LOOKUP_TABLE_PT100_SHORT` has been introduced for low memory devices? In these cases, it may be good to simply use linear approximation, according to this image, for common temperature it's a very good approximation.

<img width="236" alt="Schermata 2021-03-23 alle 15 10 19" src="https://user-images.githubusercontent.com/6470319/112160929-ee352300-8bea-11eb-9bc9-da54f20f2fee.png">

RCasatta commented on src/lib.rs:
```diff
@@ -78,102 +75,114 @@ where
-    pub fn set_calibration(&mut self, calib : u32) -> Result<(), E> {
+    ///
+    /// You can perform calibration by putting the sensor in boiling (100
+    /// degrees Celcius) water and then measuring the raw value using
```
I think Celcius is a recurring typo, it should probably be Celsius with an S?

rudihorn replied:
Indeed. I think I have now fixed all occurrences of this.

Commit:
lots of refactoring, get examples to work

Commit:
add read_ohms

rudihorn commented:
So my idea was actually for this crate to be more of a low level hardware crate that can read out the raw `max31865` outputs, leaving it up to the user and/or other crates to perform the temperature conversion. I've just added a `read_ohms` function. So if anything the temperature conversion options are more for convenience and the user can perform something like:

```  
  let raw = max31865.read_ohms().unwrap();
  let temp = temp_conversion::LOOKUP_VEC_PT1000.lookup_temperature(ohms as i32);
```

I agree a linear conversion would be nice. I will try to add this.

I've fixed up all the examples so they compile and use the newest crates etc., but still have issues with conditional dependencies for the raspberry pi example. I'm not entirely sure how best to proceed here though...

rudihorn commented:
I'm also thinking it would be nice for the code to generate those tables to be in a separate project rather than a test as is currently the case. Will do this at some stage.

RCasatta commented on .cargo/config:
```diff
@@ -1,30 +1,31 @@
 
 [target.thumbv7em-none-eabi]
-runner = 'arm-none-eabi-gdb'
+runner = 'gdb-multiarch'
```
this is a small issue on mac where there is `arm-none-eabi-gdb` but not `gdb-multiarch`

rudihorn replied:
This is a general issue between platforms. I'm using Ubuntu and they only ship a package with `arm-none-eabi-gdb` to the best of my knowledge. Using `gdb-multiarch` also seems preferable to me, so I will leave this as is for until the wider issue is resolved.

RCasatta commented on examples/rpi.rs:
```diff
@@ -11,46 +11,29 @@
-extern crate linux_embedded_hal as lin_hal;
 extern crate embedded_hal as hal;
+extern crate linux_embedded_hal as lin_hal;
 extern crate max31865;
```
what do you think to use rust edition 2018 and get rid of the `extern crate` instructions ?

rudihorn replied:
Done

RCasatta commented on src/lib.rs:
```diff
@@ -78,117 +79,139 @@ where
+        B: Unsize<[u8]>,
     {
-        let mut buffer: B  = unsafe { mem::zeroed() };
+        let mut buffer: B = unsafe { mem::zeroed() };
```
`fn read_many` is not pub and used only in a place to read `[u8; 2]`, what do you think of removing the generics and just read 2 bytes, this would allow removing the `unsafe` and also the `#![feature(unsize)]`, not requiring nightly anymore

rudihorn replied:
Done

Commit:
address review comments

Commit:
remove old readme todos

RCasatta commented:
any update?

rudihorn commented:
Sorry about the delay. I've just merged this in, pushed the changes to the main branch and have submitted it to crates as version `0.1.1`. Do let me know if there are any other further issues!
